controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+		log.Info("Waiting for IPAddress to be deleted before deleting OpenStackFloatingIPPool")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if controllerutil.RemoveFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+		log.Info("Removing finalizer from OpenStackFloatingIPPool")
+		return ctrl.Result{Requeue: true}, r.Client.Update(context.Background(), pool)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) setIPStatuses(ctx context.Context, scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool) error {
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return err
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	floatingIPs, err := networkingService.GetAllFloatingIPs()
+	if err != nil {
+		return err
+	}
+
+	pool.Status.AvailableIPs = []string{}
+	pool.Status.ClaimedIPs = []string{}
+
+	// Get claimedIPs from IPAddress
+	for _, ip := range ipAddresses.Items {
+		pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ip.Spec.Address)
+	}
+
+	for _, ip := range pool.Spec.PreAllocatedFloatingIPs {
+		if !contains(pool.Status.IPs, ip) {
+			pool.Status.IPs = append(pool.Status.IPs, ip)
+		}
+	}
+
+	ips := []string{}
+	for _, ip := range pool.Status.IPs {
+		if !contains(floatingIPs, ip) {
+			scope.Logger().Info("Floating IP not found in OpenStack, removing .Status.IPs", "ip", ip)
+			continue
+		}
+		ips = append(ips, ip)
+	}
+	pool.Status.IPs = ips
+
+	for _, ip := range pool.Status.IPs {
+		if !contains(pool.Status.ClaimedIPs, ip) {
+			pool.Status.AvailableIPs = append(pool.Status.AvailableIPs, ip)
+		}
+	}
+
+	return r.Client.Status().Update(ctx, pool)
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) getIP(ctx context.Context, scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool, openStackCluster *infrav1.OpenStackCluster, clusterName string) (string, error) {
+	if len(pool.Status.AvailableIPs) > 0 {
+		ip := pool.Status.AvailableIPs[0]
+		pool.Status.AvailableIPs = pool.Status.AvailableIPs[1:]
+		pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ip)
+		if err := r.Client.Status().Update(ctx, pool); err != nil {
+			return "", err
+		}
+		return ip, nil
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		scope.Logger().Error(err, "Failed to create networking service") // TODO Remove log
+		return "", err
+	}
+
+	fp, err := networkingService.GetOrCreateFloatingIP(pool, openStackCluster, clusterName, "")
+	if err != nil {
+		return "", fmt.Errorf("get or create floating IP: %w", err)
+	}
+	ip := fp.FloatingIP
+
+	// TODO: setStatus should probably solve this stuff
+	pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ip)
+	pool.Status.IPs = append(pool.Status.IPs, ip)
+	if err := r.Client.Status().Update(ctx, pool); err != nil {
```

Comment:
Again, we don't want to be doing mid-reconcile updates. I appreciate the desire to write our allocated state immediately, but we will sanity check this in setIPStatuses() at the start of the next reconcile anyway, so if we did crash without writing status we'd still be safe.